### PR TITLE
Alert Definitions: add hash_expression

### PR DESCRIPTION
**Pull request by zakiva**
Depends on: https://github.com/ManageIQ/manageiq/pull/16397

Changes included: 

* Currently, it is possible to create an alert definition via API only with an expression as MiqExpression. Following the changes in https://github.com/ManageIQ/manageiq/pull/15315, this PR adds the ability to create an alert with hash expression. 

* Backward Compatibility: we would like to keep the option to create an alert with expression as MiqExpression without explicitly passing a `miq_expression` attribute. So the user will have now 3 options when sending a create request and will need to pass exactly one of the following:
1. `expression` which will be saved as MiqExpression (same as before).
2. `miq_expression`.
3. `hash_expression`.  

* Allowing editing an alert definition (with similar logic). 

* Deep symbolize for data. 

(This PR replaces https://github.com/ManageIQ/manageiq/pull/14979 with different implementation)

### CODE REVIEW

**Comment by zakiva** [comment]
@miq-bot add_label enhancement

**Comment by zakiva** [comment]
@imtayadeway @moolitayer Please review
cc @simon3z @ilackarms

File: app/controllers/api/alert_definitions_controller.rb [hunk 1]
```diff
@@ -1,25 +1,39 @@
...
     end
 
     def edit_resource(type, id = nil, data = {})
       raise BadRequestError, "Must specify an id for editing a #{type} resource" unless id
+      exp_values = expression_values(data)
+      if exp_values.size > 1
+        raise BadRequestError, "Must not specify more than one expression for editing a #{type} resource"
+      elsif exp_values.size == 1
+        if data["hash_expression"]
+          data["miq_expression"] = nil
```
Comment (moolitayer) [inline app/controllers/api/alert_definitions_controller.rb]: So as we discussed with @zakiva this handles an edge case where a user wants to replace a hash expression with a miq expression or vice versa. Honestly I don't know if we should support that

**Comment by moolitayer** [comment]
@imtayadeway @abellotti We need an API review here.

**Comment by ilackarms** [comment]
@zakiva can you provide an example of a `curl` / HTTP request to add an alert definition?

**Comment by zakiva** [comment]
> @zakiva can you provide an example of a curl / HTTP request to add an alert definition?

`curl -X POST -H "Content-Type: application/json" -d "@alert.json" http://localhost:3000/api/alert_definitions --user admin:smartvm`

example of `alert.json` file: 

```
{
   "action" : "create",
   "resource" : {
   	"description" : "Alert Description",
   	"db": "ContainerNode",
   	"miq_expression": {
   		"mode": "internal",
   		"options": {
   			},
   		"eval_method": "hostd_log_threshold"
   	},
   	"options": {
   		"notifications": {
   			"email": {
   				"from": "",
                                "to": []
            }
          }
        }
   }
}
```

**Comment by zakiva** [comment]
ping @imtayadeway Can you please review?

**Comment by zakiva** [comment]
ping @abellotti @imtayadeway

File: app/controllers/api/alert_definitions_controller.rb [hunk 2]
```diff
@@ -1,26 +1,39 @@
...
   class AlertDefinitionsController < BaseController
-    REQUIRED_FIELDS = %w(description db expression options).freeze
+    REQUIRED_FIELDS = %w(description db options).freeze
 
     before_action :set_additional_attributes
 
     def create_resource(type, id, data = {})
       assert_id_not_specified(data, type)
       assert_all_required_fields_exists(data, type, REQUIRED_FIELDS)
+      if expression_values(data).size != 1
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: This looks like business logic to me. What's more - it's duplicated on line 26. This should be added to the model as a validation, which will greatly simplify this code. The code here should be almost entirely concerned with deserializing the JSON request body and passing data to the model.

File: app/controllers/api/alert_definitions_controller.rb [hunk 3]
```diff
@@ -1,26 +1,39 @@
 module Api
   class AlertDefinitionsController < BaseController
-    REQUIRED_FIELDS = %w(description db expression options).freeze
+    REQUIRED_FIELDS = %w(description db options).freeze
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: I know that you're only changing this part - but I'm also confused as to why these "required fields" aren't expressed in the model too.

Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: Agreed, I can remove them in this PR and then add the validations in the model.

File: app/controllers/api/alert_definitions_controller.rb (hunk 2 above)
Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: removed that, will add a validation in the model later on.

**Comment by zakiva** [comment]
@imtayadeway PTAL

File: app/controllers/api/alert_definitions_controller.rb [hunk 4]
```diff
@@ -1,26 +1,31 @@
 module Api
   class AlertDefinitionsController < BaseController
-    REQUIRED_FIELDS = %w(description db expression options).freeze
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: You can nix this empty line too 😁

File: app/controllers/api/alert_definitions_controller.rb [hunk 5]
```diff
@@ -1,26 +1,31 @@
...
 
     before_action :set_additional_attributes
 
     def create_resource(type, id, data = {})
       assert_id_not_specified(data, type)
-      assert_all_required_fields_exists(data, type, REQUIRED_FIELDS)
       begin
-        data["expression"] = MiqExpression.new(data["expression"])
+        create_miq_expression(data)
         data["enabled"] = true if data["enabled"].nil?
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: Again, I know you're not touching this here (and feel free to leave for a follow up if appropriate) but this is more business logic - "an alert definition is enabled by default" - that belongs in the model

File: app/controllers/api/alert_definitions_controller.rb [hunk 6]
```diff
@@ -31,5 +36,14 @@ def edit_resource(type, id = nil, data = {})
     def set_additional_attributes
       @additional_attributes = %w(expression)
     end
+
+    def expression_values(data)
+      data.slice("expression", "miq_expression", "hash_expression").values.compact
+    end
+
+    def create_miq_expression(data)
+      data["expression"] = MiqExpression.new(data["expression"]) if data["expression"]
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: I'm not sure I understand why this is necessary. Type coercion should be handled by the setter in the model - if that's not currently the case then that needs to be added so we don't have to munge the data here (and anywhere else for that matter).

File: app/controllers/api/alert_definitions_controller.rb [hunk 7]
```diff
@@ -1,26 +1,31 @@
...
-        super(type, id, data).serializable_hash.merge("expression" => data["expression"])
+        super(type, id, data.deep_symbolize_keys).serializable_hash.merge("expression" => expression_values(data).first)
       rescue => err
         raise BadRequestError, "Failed to create a new alert definition - #{err}"
       end
     end
 
     def edit_resource(type, id = nil, data = {})
       raise BadRequestError, "Must specify an id for editing a #{type} resource" unless id
+      if expression_values(data).size == 1
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: Is this still needed? Again, this is another business rule that needs to be encapsulated in the model.

File: app/controllers/api/alert_definitions_controller.rb [hunk 8]
```diff
@@ -1,26 +1,31 @@
...
 
     def create_resource(type, id, data = {})
       assert_id_not_specified(data, type)
-      assert_all_required_fields_exists(data, type, REQUIRED_FIELDS)
       begin
-        data["expression"] = MiqExpression.new(data["expression"])
+        create_miq_expression(data)
         data["enabled"] = true if data["enabled"].nil?
-        super(type, id, data).serializable_hash.merge("expression" => data["expression"])
+        super(type, id, data.deep_symbolize_keys).serializable_hash.merge("expression" => expression_values(data).first)
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: Why is the `deep_symbolize_keys` needed? If it's for a particular serialized column that is keyed with symbols, I think it would be better to make that explicit rather than converting everything wholesale

File: app/controllers/api/alert_definitions_controller.rb (hunk 4 above)
Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: done :-)

File: app/controllers/api/alert_definitions_controller.rb (hunk 5 above)
Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: Right, removing that too. This should be handled explicitly by the user.
cc @moolitayer @elad661

File: app/controllers/api/alert_definitions_controller.rb (hunk 8 above)
Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: We want the created alerts via API to be identical to alerts created in other places. 
It is not a particular (single) column, so I think it will be nicer not to list them.

File: app/controllers/api/alert_definitions_controller.rb (hunk 6 above)
Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: OK, moved it to the model: https://github.com/ManageIQ/manageiq/pull/16397

File: app/controllers/api/alert_definitions_controller.rb (hunk 7 above)
Comment (zakiva) [reply app/controllers/api/alert_definitions_controller.rb]: I will remove that too.
The intention here was to handle the case of replacing a miq expression with hash expression (and vice versa) when editing an alert, but we can leave that to the user to explicitly nullify these fields. 
cc @moolitayer @elad661

**Comment by zakiva** [comment]
@imtayadeway Thanks for the review! I made the requested changes.

**Comment by elad661** [comment]
@imtayadeway now that the change in the model ( https://github.com/ManageIQ/manageiq/pull/16397 ) is merged, can we get this PR merged too, or are there any other required changes?

**Comment by bazulay** [comment]
@imtayadeway @abellotti  PTAL

**Comment by elad661** [comment]
@imtayadeway Can you please take a look? This PR has been waiting for a re-review for almost a month now...

**Comment by bazulay** [comment]
@gtanzillo  ping

File: app/controllers/api/alert_definitions_controller.rb [hunk 9]
```diff
@@ -31,5 +28,10 @@ def edit_resource(type, id = nil, data = {})
     def set_additional_attributes
       @additional_attributes = %w(expression)
     end
+
+    def update_miq_expression(data)
+      data["miq_expression"] = data["expression"]
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: I'm still not sure why this is needed. Doesn't the model support this via both `#expression=` and `#miq_expression=` ?

File: app/controllers/api/alert_definitions_controller.rb [hunk 10]
```diff
@@ -1,16 +1,13 @@
...
 
     def create_resource(type, id, data = {})
       assert_id_not_specified(data, type)
-      assert_all_required_fields_exists(data, type, REQUIRED_FIELDS)
       begin
-        data["expression"] = MiqExpression.new(data["expression"])
-        data["enabled"] = true if data["enabled"].nil?
-        super(type, id, data).serializable_hash.merge("expression" => data["expression"])
+        update_miq_expression(data) if data["expression"]
+        alert = super(type, id, data.deep_symbolize_keys).serializable_hash
```
Comment (imtayadeway) [inline app/controllers/api/alert_definitions_controller.rb]: I'm concerned that this `deep_symbolize_keys` will corrupt an incoming `MiqExpression`:

```ruby
{"=" => {"field" => "foo", "value" => "bar"}}.deep_symbolize_keys # => {:"=" => {:field => "foo", :value => "bar"}}
```

File: spec/requests/alert_definitions_spec.rb [hunk 11]
```diff
@@ -81,7 +81,7 @@
     alert_definition = MiqAlert.find(response.parsed_body["results"].first["id"])
     expect(alert_definition).to be_truthy
     expect(alert_definition.expression.class).to eq(MiqExpression)
-    expect(alert_definition.expression.exp).to eq(sample_alert_definition["expression"])
+    expect(alert_definition.expression.exp).to eq(sample_alert_definition["expression"].deep_symbolize_keys)
```
Comment (imtayadeway) [inline spec/requests/alert_definitions_spec.rb]: This looks wrong, because miq expressions are not keyed with symbols (see above)

File: spec/requests/alert_definitions_spec.rb [hunk 12]
```diff
@@ -91,6 +91,66 @@
...
+      "options"        => { "notifications" => {"delay_next_evaluation" => 600, "evm_event" => {} } },
+      "enabled"        => true
+    }
+    api_basic_authorize collection_action_identifier(:alert_definitions, :create)
+    post(api_alert_definitions_url, :params => sample_alert_definition)
+    expect(response).to have_http_status(:ok)
+    alert_definition = MiqAlert.find(response.parsed_body["results"].first["id"])
+    expect(alert_definition).to be_truthy
+    expect(alert_definition.expression.class).to eq(MiqExpression)
+    expect(alert_definition.expression.exp).to eq(sample_alert_definition["miq_expression"].deep_symbolize_keys)
```
Comment (imtayadeway) [inline spec/requests/alert_definitions_spec.rb]: As above

File: spec/requests/alert_definitions_spec.rb [hunk 13]
```diff
@@ -91,6 +91,66 @@
     )
   end
 
+  it "creates an alert definition with miq_expression" do
+    sample_alert_definition = {
+      "description"    => "Test Alert Definition",
+      "db"             => "ContainerNode",
+      "miq_expression" => { "eval_method" => "dwh_generic", "mode" => "internal", "options" => {} },
```
Comment (imtayadeway) [inline spec/requests/alert_definitions_spec.rb]: Is this supposed to be a hash expression standing in place of an miq expression? Even if this "works", I wouldn't want us to officially support this surprising behavior

File: spec/requests/alert_definitions_spec.rb [hunk 14]
```diff
@@ -91,6 +91,66 @@
...
+      "expression"  => sample_alert_definition["hash_expression"]
+    )
+  end
+
+  it "fails to create an alert definition with more than one expression" do
+    sample_alert_definition = {
+      "description"     => "Test Alert Definition",
+      "db"              => "ContainerNode",
+      "hash_expression" => { "eval_method" => "nothing", "mode" => "internal", "options" => {} },
+      "miq_expression"  => { "eval_method" => "dwh_generic", "mode" => "internal", "options" => {} },
```
Comment (imtayadeway) [inline spec/requests/alert_definitions_spec.rb]: As above

File: spec/requests/alert_definitions_spec.rb [hunk 15]
```diff
@@ -156,6 +216,77 @@
...
+      :options         => { :notifications => {:delay_next_evaluation => 0, :evm_event => {} } }
+    )
+
+    exp = { :eval_method => "mw_heap_used", :mode => "internal", :options => {} }
+
+    post(
+      api_alert_definition_url(nil, alert_definition),
+      :params => {
+        :action          => "edit",
+        :miq_expression  => exp,
```
Comment (imtayadeway) [inline spec/requests/alert_definitions_spec.rb]: As above

File: spec/requests/alert_definitions_spec.rb [hunk 16]
```diff
@@ -156,6 +216,77 @@
...
+      :hash_expression => { :eval_method => "nothing", :mode => "internal", :options => {} },
+      :options         => { :notifications => {:delay_next_evaluation => 0, :evm_event => {} } }
+    )
+
+    post(
+      api_alert_definition_url(nil, alert_definition),
+      :params => {
+        :action          => "edit",
+        :hash_expression => { :eval_method => "event_threshold", :mode => "internal", :options => {} },
+        :miq_expression  => { :eval_method => "mw_heap_used", :mode => "internal", :options => {} }
```
Comment (imtayadeway) [inline spec/requests/alert_definitions_spec.rb]: As above

File: app/controllers/api/alert_definitions_controller.rb (hunk 9 above)
Comment (elad661) [reply app/controllers/api/alert_definitions_controller.rb]: This is needed because the setter in the model for `expression=` looks like this: https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_alert.rb#L59-L65
```ruby
    if exp.kind_of?(MiqExpression)
      self.miq_expression = exp
    elsif exp.kind_of?(Hash)
      self.hash_expression = exp
```

And the data we get from the api user is always a "hash" (but not a Hash Expression), because it's parsed json.

The model will only try to create a new MiqExpression from hash data if you explicitly use the `miq_expression` setter, so we need to check if the user provided us an "expression" parameter, and if so *explicitly* say we want a miq_expression and not an hash expression.

Since hash expressions are just ruby hashes, there's no cleaner way to do this - I don't think we'd want the model to attempt parsing hashes to see if they fit the "miq expression" or "hash expression" forms and magically decide which type to use.

**Comment by elad661** [comment]
Thanks for the review. I fixed the tests and changed the controller to only do `.deep_symbolize_keys` on hash expressions (since those do need to be symbolized).

File: app/controllers/api/alert_definitions_controller.rb (hunk 9 above)
Comment (imtayadeway) [reply app/controllers/api/alert_definitions_controller.rb]: This is OK for now....can we look at deprecating the old API (i.e. `MiqAlert#expression`) in a follow up PR? Then I think a lot of the complexity of this controller could go away.

File: spec/requests/alert_definitions_spec.rb (hunk 16 above)
Comment (imtayadeway) [reply spec/requests/alert_definitions_spec.rb]: It looks like this is still a "hash_expression" in place of an "miq_expression"?

Comment (elad661) [reply spec/requests/alert_definitions_spec.rb]: oh, you're right, I didn't actually fix this one when I said it was. I fixed it now.

File: app/controllers/api/alert_definitions_controller.rb (hunk 9 above)
Comment (elad661) [reply app/controllers/api/alert_definitions_controller.rb]: I'm fine with deprecating the `MiqAlert#expression`, and I can do a follow-up PR.

@moolitayer is there any reason to keep the old `MiqAlert#expression` API around?

Comment (moolitayer) [reply app/controllers/api/alert_definitions_controller.rb]: In the model we would need to keep the `expression` due to the alerts screen.
In the API I agree we should deprecate and then drop.
(deprecate in Gaprindashvili and Drop in H?)

File: spec/requests/alert_definitions_spec.rb [hunk 17]
```diff
@@ -71,7 +71,7 @@
     sample_alert_definition = {
       "description" => "Test Alert Definition",
       "db"          => "ContainerNode",
-      "expression"  => { "eval_method" => "dwh_generic", "mode" => "internal", "options" => {} },
+      "expression"  => {"exp" => {"=" => {"field" => "Vm-name", "value" => "foo"}}, "context_type" => nil},
```
Comment (abellotti) [inline spec/requests/alert_definitions_spec.rb]: Just curious about these changes, description mentioned being backward compatible, why are these needed ?

Comment (elad661) [reply spec/requests/alert_definitions_spec.rb]: Because, as @imtayadeway noted in this comment https://github.com/ManageIQ/manageiq-api/pull/76#discussion_r154723513, the expression there was not actually a MiqExpression, so that means the test was wrong. It wasn't supposed to "work" like that.

Comment (abellotti) [reply spec/requests/alert_definitions_spec.rb]: 👍 thanks for clarifying.

**Comment by abellotti** [comment]
Thanks @zakiva for the API Enhancement and everyone for the reviews.  This LGTM!! 👍
